source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -24,11 +25,113 @@ void MapdataCheckPoint::read(EGG::Stream &stream) {
+    if (m_prevPt == 0xff) {
+        MapdataCheckPath *checkpath = courseMap->checkPath()->findCheckpathForCheckpoint(id);
+        m_prevCount = 0;
+
+        for (size_t i = 0; i < 6; i++) {
+            u16 prevID = checkpath->getPrev(i);
+
+            if (prevID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *prev = courseMap->checkPath()->get(prevID);
+            m_prevPoints[i] = accessor.get(prev->end());
+            m_prevCount++;
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        m_prevCount++;
+    }
+
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == 0xff) {
+        MapdataCheckPath *checkpath = courseMap->checkPath()->findCheckpathForCheckpoint(id);
+        m_nextCount = 0;
+
+        for (size_t i = 0; i < 6; i++) {
+            u16 nextID = checkpath->getNext(i);
+
+            if (nextID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *next = courseMap->checkPath()->get(nextID);
+            m_nextPoints[i].checkpoint = accessor.get(next->start());
+            m_nextCount++;
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        m_nextCount++;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (size_t i = 0; i < 6; i++) {
+        if (i < m_nextCount) {
+            auto next = nextPoint(i);
+            m_nextPoints[i].distance = (next->m_midpoint - m_midpoint).normalise();
+            m_nextPoints[i].p0diff =
+                    EGG::Vector2f(next->m_left.x - m_left.x, next->m_left.y - m_left.y);
+            m_nextPoints[i].p1diff =
+                    EGG::Vector2f(next->m_right.x - m_right.x, next->m_right.y - m_right.y);
+        } // else mkw initalizes to zero (pointless) at 0x80515a20
+        // i can't prove it's pointless, but if the PR reviewer would kindly please advise, thank
+        // you i didn't bytematch this btw but i tried a little and gave up
+        // https://decomp.me/scratch/jYvUa
+    }
+}
+
+s8 MapdataCheckPoint::checkArea() const {
+    return m_checkArea;
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(size_t i) const {
+    ASSERT(i < m_nextPoints.size());
+    return m_nextPoints[i].checkpoint;
+}
+
+/// @brief Finds indices of the finish line and last key checkpoint; also calls initCheckpointLinks
+/// for all checkpoints
+/// @addr{0x80515244}
+void MapdataCheckPointAccessor::init() {
+    s8 lastKcpType = -1;
+    s16 finishLineCheckpointId = -1;
+
+    for (size_t ckptId = 0; ckptId < size(); ckptId++) {
+        MapdataCheckPoint *checkpoint = get(ckptId);
+        checkpoint->initCheckpointLinks(*this, ckptId);
+        checkpoint = get(ckptId);
+
+        if (checkpoint->checkArea() == MapdataCheckPoint::FINISH_LINE) {
```

Comment:
You use `checkArea` twice, can we add a new variable to store this to avoid two calls to the getter function